app/controllers/user_subscriptions_controller.rb

```diff
@@ -0,0 +1,87 @@
+class UserSubscriptionsController < ApplicationController
+  before_action :authenticate_user!
+
+  def subscribed
+    params.require(%i[source_type source_id])
+    source_type = params[:source_type]
+    source_id = params[:source_id]
+
+    is_subscribed = UserSubscriptions::CacheChecker.new(current_user, source_type, source_id).cached_subscription_check
+
+    render json: {
+      is_subscribed: is_subscribed,
+      status: 200
+    }, status: :ok
+  end
+
+  def create
```

Comment:
This method is quite long. Feel free to leave as is, but a pattern I commonly follow for this is a service object like `UserSubscriptions::Create`, which returns an object (an `OpenStruct` most likely) that responds to `success?`, `error` (for when `success?` is `false`) and `data` (when `success?` is true). Then this whole action would look something like:

```ruby
def create
  user_subscription = UserSubscriptions::Create.call(...)

  if user_subsription.success?
    rate_limiter.track_limit_by_action(:user_subscription_creation)
    render json: { message: "success", status: 200 }, status: :ok
  else
    error_response(user_subscription.error)
  end
end
```

Reply:
I love this idea! I [marked it down](https://github.com/thepracticaldev/dev.to/projects/19#card-40564468) to circle back in another PR. I suspect this controller will be tweaked further in the next PR so it will be even easier to set up a new service once those final details are known 💯